File: rhizome/security/rate_limiter.py
"""
Rate limiting для защиты от атак
"""

import time
from collections import defaultdict, deque
from typing import Dict, Optional, Tuple

from rhizome.exceptions import RateLimitError
from rhizome.logger import get_logger


class RateLimiter:
    """Rate limiter для ограничения частоты запросов"""

    def __init__(self, max_requests: int = 100, window_seconds: int = 60, per_node_limit: int = 20):
        """
        Инициализация rate limiter

        Args:
            max_requests: Максимальное количество запросов в окне
            window_seconds: Размер окна в секундах
            per_node_limit: Лимит запросов от одного узла в окне
        """
        self.max_requests = max_requests
        self.window_seconds = window_seconds
        self.per_node_limit = per_node_limit

        # История запросов: (timestamp, node_id)
        self.request_history: deque = deque(maxlen=max_requests * 2)

        # Запросы по узлам
        self.node_requests: Dict[bytes, deque] = defaultdict(
            lambda: deque(maxlen=per_node_limit * 2)
        )

        self.logger = get_logger("security.rate_limiter")

    def check_rate_limit(self, node_id: Optional[bytes] = None) -> bool:
        """
        Проверка rate limit

        Args:
            node_id: ID узла (опционально, для per-node лимита)

        Returns:
            True если запрос разрешен, False если превышен лимит

        Raises:
            RateLimitError если превышен лимит
        """
        current_time = time.time()

        # Очищаем старые запросы
        self._cleanup_old_requests(current_time)

        # Проверяем общий лимит
        recent_requests = sum(
            1
            for timestamp in self.request_history
            if current_time - timestamp < self.window_seconds
        )

        if recent_requests >= self.max_requests:
            self.logger.warning(
                "Rate limit exceeded", requests=recent_requests, limit=self.max_requests
            )
            raise RateLimitError(
                f"Rate limit exceeded: {recent_requests}/{self.max_requests} requests in {self.window_seconds}s"
            )

        # Проверяем per-node лимит
        if node_id:
            node_recent = sum(
                1
                for timestamp in self.node_requests[node_id]
                if current_time - timestamp < self.window_seconds
            )

            if node_recent >= self.per_node_limit:
                self.logger.warning(
                    "Per-node rate limit exceeded",
                    node_id=node_id.hex()[:16],
                    requests=node_recent,
                    limit=self.per_node_limit,
                )
                raise RateLimitError(
                    f"Per-node rate limit exceeded: {node_recent}/{self.per_node_limit} requests"
                )

        # Запрос разрешен, добавляем в историю
        self.request_history.append(current_time)
        if node_id:
            self.node_requests[node_id].append(current_time)

        return True

    def _cleanup_old_requests(self, current_time: float):
        """Очистка старых запросов из истории"""
        # Очищаем общую историю
        while (
            self.request_history and (current_time - self.request_history[0]) > self.window_seconds
        ):
            self.request_history.popleft()

        # Очищаем историю по узлам
        for node_id in list(self.node_requests.keys()):
            node_history = self.node_requests[node_id]
            while node_history and (current_time - node_history[0]) > self.window_seconds:
                node_history.popleft()

            # Удаляем пустые истории
            if not node_history:
                del self.node_requests[node_id]

    def get_stats(self) -> Dict[str, int]:
        """Получение статистики rate limiter"""
        current_time = time.time()
        self._cleanup_old_requests(current_time)

        recent_requests = sum(
            1
            for timestamp in self.request_history
            if current_time - timestamp < self.window_seconds
        )

        return {
            "recent_requests": recent_requests,
            "max_requests": self.max_requests,
            "window_seconds": self.window_seconds,
            "active_nodes": len(self.node_requests),
        }

File: src/security/rate_limiter.rs
use std::collections::{HashMap, VecDeque};
use tracing::warn;

use crate::exceptions::{NetworkError, RhizomeError};
use crate::utils::time::get_now_f64;

/// Structure for limit messages peer some period of time
///
/// Use algo of Sliding Window
pub struct RateLimiter {
    /// Global limit of requests
    max_requests: usize,
    /// Size of sliding window
    window_seconds: u64,
    /// Limit for one node
    per_node_limit: usize,

    /// History of all requests (timestamps)
    request_history: VecDeque<f64>,

    /// Requests by node: NodeID -> deque of timestamps
    node_requests: HashMap<Vec<u8>, VecDeque<f64>>,
}

impl RateLimiter {
    /// Initialize rate limiter
    pub fn new(max_requests: usize, window_seconds: u64, per_node_limit: usize) -> Self {
        Self {
            max_requests,
            window_seconds,
            per_node_limit,
            request_history: VecDeque::with_capacity(max_requests * 2),
            node_requests: HashMap::new(),
        }
    }

    /// Check rate limit
    ///
    /// Main function which work with all requests and can block some requests if they do not fit
    pub fn check_rate_limit(&mut self, node_id: Option<&[u8]>) -> Result<bool, RhizomeError> {
        let current_time = get_now_f64();

        self.cleanup_old_requests(current_time);

        let recent_requests = self.request_history.len();

        if recent_requests >= self.max_requests {
            warn!(
                requests = recent_requests,
                limit = self.max_requests,
                "Rate limit exceeded"
            );
            return Err(RhizomeError::Network(NetworkError::RateLimitError));
        }

        if let Some(id) = node_id {
            let node_id_vec = id.to_vec();
            let node_history = self
                .node_requests
                .entry(node_id_vec.clone())
                .or_insert_with(|| VecDeque::with_capacity(self.per_node_limit * 2));

            let node_recent = node_history.len();

            if node_recent >= self.per_node_limit {
                let hex_id = hex::encode(&id[..id.len().min(8)]);
                warn!(
                    node_id = %hex_id,
                    requests = node_recent,
                    limit = self.per_node_limit,
                    "Per-node rate limit exceeded"
                );
                return Err(RhizomeError::Network(NetworkError::RateLimitError));
            }

            node_history.push_back(current_time);
        }

        self.request_history.push_back(current_time);

        Ok(true)
    }

    /// Cleanup old requests by window size in sliding window
    fn cleanup_old_requests(&mut self, current_time: f64) {
        let window = self.window_seconds as f64;

        while let Some(&first_ts) = self.request_history.front() {
            if current_time - first_ts > window {
                self.request_history.pop_front();
            } else {
                break;
            }
        }

        self.node_requests.retain(|_, history| {
            while let Some(&first_ts) = history.front() {
                if current_time - first_ts > window {
                    history.pop_front();
                } else {
                    break;
                }
            }

            !history.is_empty()
        });
    }

    /// Getting requests statistics for analyze
    pub fn get_stats(&mut self) -> HashMap<String, f64> {
        let current_time = get_now_f64();
        self.cleanup_old_requests(current_time);

        let mut stats = HashMap::new();
        stats.insert(
            "recent_requests".to_string(),
            self.request_history.len() as f64,
        );
        stats.insert("max_requests".to_string(), self.max_requests as f64);
        stats.insert("window_seconds".to_string(), self.window_seconds as f64);
        stats.insert("active_nodes".to_string(), self.node_requests.len() as f64);

        stats
    }
}
